File: analytics/schema/python/dora_event.py
from __future__ import annotations

from datetime import UTC, datetime
from typing import Literal, Optional, Union
from uuid import uuid4

from pydantic import BaseModel, ConfigDict, Field

DoraStatus = Literal["success", "failure"]
DoraEnvironment = Literal["sandbox", "staging", "production", "local"]
DoraStage = Literal[
    "init",
    "branch-create",
    "check",
    "pr-pipeline",
    "deploy-sandbox",
    "deploy-staging",
    "deploy-production",
    "integration-pipeline",
    "collect",
    "analyze",
]


class DoraEventV1(BaseModel):
    version: Literal["1.0"] = "1.0"
    work_id: str
    team: str
    stack: str
    stage: DoraStage
    environment: DoraEnvironment = "local"
    status: DoraStatus
    duration_ms: int = 0
    timestamp: str = Field(
        default_factory=lambda: datetime.now(UTC).isoformat().replace("+00:00", "Z")
    )

    model_config = ConfigDict(extra="forbid")


class DoraEventV2(DoraEventV1):
    version: Literal["2.0"] = "2.0"
    event_id: str = Field(default_factory=lambda: str(uuid4()))
    correlation_id: Optional[str] = None
    repo: Optional[str] = None
    service: Optional[str] = None
    commit_sha: Optional[str] = None
    pr_number: Optional[int] = None
    workflow_run_id: Optional[str] = None
    actor: Optional[str] = None
    failure_reason: Optional[str] = None

    model_config = ConfigDict(extra="forbid")


DoraEvent = Union[DoraEventV1, DoraEventV2]


def parse_event(data: dict) -> DoraEvent:
    """Parse a dict into the correct DoraEvent version."""
    version = data.get("version", "1.0")
    if version == "2.0":
        return DoraEventV2(**data)
    return DoraEventV1(**data)


def emit_event(event: DoraEvent) -> None:
    """Emit event as single-line JSON to stdout."""
    print(event.model_dump_json())

File: analytics/schema/typescript/src/dora-event.ts
import { z } from "zod";

export const DoraStatusSchema = z.enum(["success", "failure"]);
export const DoraEnvironmentSchema = z.enum([
  "sandbox",
  "staging",
  "production",
  "local",
]);
export const DoraStageSchema = z.enum([
  "init",
  "branch-create",
  "check",
  "pr-pipeline",
  "deploy-sandbox",
  "deploy-staging",
  "deploy-production",
  "integration-pipeline",
  "collect",
  "analyze",
]);

export const DoraEventV1Schema = z.object({
  version: z.literal("1.0"),
  work_id: z.string(),
  team: z.string(),
  stack: z.string(),
  stage: DoraStageSchema,
  environment: DoraEnvironmentSchema.default("local"),
  status: DoraStatusSchema,
  duration_ms: z.number().int().default(0),
  timestamp: z.string(),
});

export const DoraEventV2Schema = DoraEventV1Schema.extend({
  version: z.literal("2.0"),
  event_id: z.string().uuid(),
  correlation_id: z.string().optional(),
  repo: z.string().optional(),
  service: z.string().optional(),
  commit_sha: z.string().optional(),
  pr_number: z.number().int().positive().optional(),
  workflow_run_id: z.string().optional(),
  actor: z.string().optional(),
  failure_reason: z.string().optional(),
});

export const DoraEventSchema = z.discriminatedUnion("version", [
  DoraEventV1Schema,
  DoraEventV2Schema,
]);

export type DoraEventV1 = z.infer<typeof DoraEventV1Schema>;
export type DoraEventV2 = z.infer<typeof DoraEventV2Schema>;
export type DoraEvent = z.infer<typeof DoraEventSchema>;

export function parseEvent(data: unknown): DoraEvent {
  return DoraEventSchema.parse(data);
}
